### PR TITLE
Process 3DS modal success requests when changing a subscriptions payment method

### LINKED ISSUE #3312: Updating Subscription's payment method to a 3DS card results in an error
**Describe the bug**

Attempting to change a subscription's payment method to a 3D Secure payment method will result in the subscription being set to **Pending** and the following error message: 

<p align="center">
Screenshot 2024-07-24 at 4 51 05 PM
</p> 

**To Reproduce**
1. Enable WooCommerce Subscriptions
2. Purchase a subscription product. 
3. From the **My Account > Subscriptions** page view the subscription. 
4. Click the **Change payment** button.
5. On the change payment page enter a 3D-secure enabled card. eg `4000002760003184`.
6. The page will do a brief redirect before the 3D-Secure challenge modal will appear. 
7. Complete the auth.
8. You will then see the error message mentioned above. 
9. If you view the subscription in the Admin dashboard you will also see this note:

<p align="center">
Screenshot 2024-07-24 at 4 55 35 PM
</p> 

**Expected behavior**
There should be no error and the subscription should not be set to pending-payment.

---

It appears that after displaying the 3DS modal, the `confirmIntent()` JS function [makes an AJAX request to the `update_order_status` action](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/develop/client/api/index.js#L345-L354) and that attempts to confirm payment even though this request is for a change subscription payment method and no payment is needed.

### PR DESCRIPTION
**Pull request by james-allan**
Fixes #3312

## Changes proposed in this Pull Request:

In https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3307, I changed setup intent requests to make sure we use the Stripe SDK to complete the next actions (eg 3D-Secure challenges). This means the modal pops up rather than a redirect. 

This inadvertently broke the change payment method flow for subscriptions because after completing the 3D-Secure challenge, the JS attempts to confirm the order payment. Because the "order" in this context is actually a subscription, you would get 2 issues:

1. The intent validation would fail. 
2. The subscription would be set to "pending". 

This PR fixes that by carving out a flow for change payment method requests.

> [!note]
> https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3307 is unreleased so this error doesn't currently impact the public version of the plugin.

## Testing instructions

1. Enable WooCommerce Subscriptions
2. Purchase a subscription product.
3. From the **My Account > Subscriptions** page view the subscription.
4. Click the Change payment button.
5. On the change payment page enter a 3D-secure enabled card. eg `4000002760003184`.
6. The page will do a brief redirect before the 3D-Secure challenge modal will appear.
7. Complete the auth.
   - On `develop` you will see the following error and the subscription will be "pending". 

| Error message | Subscription "Pending" and Note |
|--------|--------|
| Screenshot 2024-07-25 at 2 11 18 PM | Screenshot 2024-07-25 at 2 11 38 PMScreenshot 2024-07-25 at 2 12 41 PM | 

8. Repeat the steps on this branch and the subscription change payment request should complete as expected. 

---

-   [ ] Covered with tests (or have a good reason not to test in description ☝️)
-   [x] Added changelog entry **in both** `changelog.txt` and `readme.txt` (or does not apply)
-   [ ] Tested on mobile (or does not apply)

**Post merge**

-   [ ] Added testing instructions to the [Release Testing Instructions wiki page](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Release-Testing-Instructions) (or does not apply)
-   [ ] Added the [needs docs label](https://github.com/woocommerce/woocommerce-gateway-stripe/labels?q=docs) (or does not apply)
-   [ ] Included this PR in the Release Thread scope (or does not apply)

### CODE REVIEW

File: includes/class-wc-stripe-intent-controller.php [hunk 1]
```diff
@@ -1037,7 +1039,57 @@ public function create_and_confirm_setup_intent_ajax() {
...
+			if ( ! $is_nonce_valid ) {
+				throw new WC_Stripe_Exception( 'missing-nonce', __( 'CSRF verification failed.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			if ( ! function_exists( 'wcs_is_subscription' ) ) {
+				throw new WC_Stripe_Exception( 'subscriptions_not_found', __( "We're not able to process this subscription change payment request payment. Please try again later.", 'woocommerce-gateway-stripe' ) );
+			}
+
+			$subscription_id = isset( $_POST['order_id'] ) ? absint( $_POST['order_id'] ) : false;
+			$subscription    = wcs_get_subscription( $subscription_id );
```
Comment (wjrosa) [inline includes/class-wc-stripe-intent-controller.php]: `$subscription_id` can be `false` above. In this case, I think we should either:
1 - Make `$subscription` `false` as well (`$subscription = $subscription_id ? wcs_get_subscription( $subscription_id ) : false;`)
2 - Validate `$subscription_id` and throw an error as we do just below for `$subscription`
3 - Only load the subscription if `$subscription_id` is not `false`

File: includes/class-wc-stripe-intent-controller.php [hunk 2]
```diff
@@ -1037,7 +1039,57 @@ public function create_and_confirm_setup_intent_ajax() {
...
+
+			if ( ! $subscription ) {
+				throw new WC_Stripe_Exception( 'subscription_not_found', __( "We're not able to process this subscription change payment request payment. Please try again later.", 'woocommerce-gateway-stripe' ) );
+			}
+
+			$setup_intent_id = isset( $_POST['intent_id'] ) ? wc_clean( wp_unslash( $_POST['intent_id'] ) ) : null;
+
+			$gateway = $this->get_upe_gateway();
+			$gateway->create_token_from_setup_intent( $setup_intent_id, $subscription->get_user() );
```
Comment (wjrosa) [inline includes/class-wc-stripe-intent-controller.php:1083]: Same case here for `$setup_intent_id`. The first param of`create_token_from_setup_intent` is not type-strict, but a bug may happen with the Stripe request (due the generated URL path):
```
			$setup_intent = $this->stripe_request( 'setup_intents/' . $setup_intent_id . '?&expand[]=latest_attempt' );
```

**Review by wjrosa — APPROVED** [review]
Code is good and works as expected:
Screenshot 2024-07-25 at 11 55 50

Left some minor comments 👍

File: includes/class-wc-stripe-intent-controller.php (hunk 1 above)
Comment (james-allan) [reply includes/class-wc-stripe-intent-controller.php]: Thanks @wjrosa. I've gone with option 1 and only call `wcs_get_subscription()` if we have an ID. [4fa2fd1](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3313/commits/4fa2fd1a8dae406af5fc469d5eec04f2825ede4e)

File: includes/class-wc-stripe-intent-controller.php (hunk 2 above)
Comment (james-allan) [reply includes/class-wc-stripe-intent-controller.php:1083]: 👍 Fixed in: [6b271c9](https://github.com/woocommerce/woocommerce-gateway-stripe/pull/3313/commits/6b271c975dda085fb15586379a8643bdc0787efc)

**Pull request merged**